sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotInstantiateSharedClasses.cs

```diff
@@ -0,0 +1,82 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotInstantiateSharedClasses : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4277";
+        private const string MessageFormat = "Refactor this code so that it doesn't invoke the constructor of this class.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    if (c.IsTest())
```

Comment:
This is a sensible exception, but it isn't listed in the RSPEC description. Should it be?

Also, any idea why the RSPEC definition exclude VB.NET? MEF rules are just as applicable there.

Reply:
@duncanp-sonar The exception is actually listed when reading the `Analysis Scope` which targets only `Main Sources`. We know this is at the moment not displayed to end-users and there are numerous MMF on platform and tooling side regarding this.

I have updated the RSPEC to also include VB.Net

Reply:
Indeed, it should listed and it is :) "See Analysis Scope: Main Sources".

Good idea to cover VB, I'll create a ticket.